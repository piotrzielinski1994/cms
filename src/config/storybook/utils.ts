import { ReactNode } from 'react';

const DEFAULT_VALUE = '__default__' as const;
const THUMBNAIL_ID = 'thumbnail';

const storyToUrlPath = (title: string) => {
  const parts = title.split('/');
  const story = parts[parts.length - 1];
  const titlePart = title.toLowerCase().replace(/\//g, '-').replace(/\s+/g, '-');
  const storyPart = story
    .replace(/([a-z0-9])([A-Z])/g, '$1-$2')
    .replace(/([a-zA-Z])(\d+)/g, '$1-$2')
    .toLowerCase();
  return `${titlePart}--${storyPart}`;
};

const getFallback = <T extends ReactNode>(value: T, translation: string): T => {
  return value === DEFAULT_VALUE ? (translation as T) : value;
};

export { DEFAULT_VALUE, getFallback, storyToUrlPath, THUMBNAIL_ID };
