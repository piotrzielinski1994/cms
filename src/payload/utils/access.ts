import { User } from '@/payload.types';
import { Access, FieldAccess } from 'payload';

const anyone: Access = () => true;

const authenticated: Access = ({ req: { user } }) => {
  return Boolean(user);
};

const admin: Access = ({ req: { user } }) => {
  return Boolean(user?.roles?.includes('admin'));
};

const adminOrSelf: Access = ({ req: { user } }) => {
  if (!user) return false;
  if (user.roles?.includes('admin')) return true;
  return { id: { equals: user.id } };
};

const adminFieldLevel: FieldAccess<{ id: string }, User> = ({ req: { user } }) => {
  return Boolean(user?.roles?.includes('admin'));
};

const authenticatedOrPublished: Access = ({ req: { user } }) => {
  if (user) return true;
  return { _status: { equals: 'published' } };
};

export { admin, adminFieldLevel, adminOrSelf, anyone, authenticated, authenticatedOrPublished };
