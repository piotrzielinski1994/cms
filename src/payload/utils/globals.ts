import type { Config } from '@/payload.types';
import configPromise from '@/payload/payload.config';
import { Locale } from 'next-intl';
import { unstable_cache } from 'next/cache';
import { getPayload } from 'payload';

const getGlobal = async (slug: keyof Config['globals'], locale: Locale = 'en', depth = 1) => {
  const payload = await getPayload({ config: configPromise });
  return payload.findGlobal({ slug, depth, locale });
};

const getCachedGlobal = (slug: keyof Config['globals'], locale: Locale, depth = 1) => {
  const cacheConfig = { tags: [`global__${locale}_${slug}`] };
  return unstable_cache(() => getGlobal(slug, locale, depth), [slug, locale], cacheConfig);
};

export { getCachedGlobal };
