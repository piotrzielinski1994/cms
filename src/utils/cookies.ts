import { setCookie } from 'typescript-cookie';

const setPermament = (name: string, value: string | number | boolean | object) => {
  setCookie(name, value.toString(), { expires: 365 * 10, path: '/' });
};

const cookies = { setPermament };

export default cookies;
