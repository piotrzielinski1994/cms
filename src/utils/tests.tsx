import { Providers } from '@/providers';
import { ComponentProps, ReactNode } from 'react';

const resolveServerComponent = async <T extends (props: unknown) => Promise<ReactNode>>(
  Component: T,
  props: Parameters<T>[0],
) => {
  const ComponentResolved = await Component(props);
  return () => ComponentResolved;
};

const withProviders = (
  children: ReactNode,
  props: Partial<Omit<ComponentProps<typeof Providers>, 'children'>> = {},
) => {
  const overwrittenProps: ComponentProps<typeof Providers> = {
    locale: 'en',
    initialTheme: 'light',
    initialFontScale: 'base',
    initialCookiesConsent: true,
    ...props,
    children,
  };
  return <Providers {...overwrittenProps} />;
};

export { resolveServerComponent, withProviders };
