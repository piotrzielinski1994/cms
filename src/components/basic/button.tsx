import { Link } from '@/config/next.routing.config';
import { cn } from '@/utils/tailwind';
import { ButtonHTMLAttributes, ComponentProps } from 'react';

type NativeButtonProps = BaseButtonProps & ButtonHTMLAttributes<HTMLButtonElement>;
type LinkButtonProps = BaseButtonProps & ComponentProps<typeof Link>;
type BaseButtonProps = {
  variant?: keyof typeof buttonClassNames.variant;
};

const buttonClassNames = {
  button: cn(
    'px-6 py-3',
    'border-2 border-primary hover:border-primary/90',
    'font-semibold',
    'disabled:cursor-not-allowed box-content',
  ),
  variant: {
    primary: cn('bg-primary hover:bg-primary/90 text-primary-foreground', 'disabled:bg-primary/50'),
    secondary: cn(
      ' text-primary',
      'disabled:border-primary/50 disabled:text-primary/50',

      'bg-transparent hover:bg-foreground/5',
    ),
  },
};

const Button = ({ className, variant = 'primary', ...rest }: NativeButtonProps) => {
  return (
    <button
      type="button"
      className={cn(buttonClassNames.button, buttonClassNames.variant[variant], className)}
      {...rest}
    />
  );
};

const ButtonLink = ({ className, variant = 'primary', ...rest }: LinkButtonProps) => {
  return (
    <Link
      className={cn(buttonClassNames.button, buttonClassNames.variant[variant], className)}
      {...rest}
    />
  );
};

export { Button, buttonClassNames, ButtonLink };
