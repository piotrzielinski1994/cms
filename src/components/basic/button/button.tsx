import { Link } from '@/config/next.routing.config';
import { EnhancedHtmlProps } from '@/utils/html/html.types';
import { cn } from '@/utils/tailwind';
import { ComponentProps, forwardRef } from 'react';

// prettier-ignore
type ButtonProps = EnhancedHtmlProps<'button', {
  variant?: keyof (typeof buttonClassNames)['variant'];
}>;

type LinkButtonProps = ComponentProps<typeof Link> & Pick<ButtonProps, 'disabled' | 'variant'>;

const buttonClassNames = {
  button: cn(
    'px-6 py-3',
    'border-2 border-primary hover:border-primary/90',
    'font-semibold whitespace-nowrap',
    'disabled:cursor-not-allowed',
    'focus-visible:tw-cms-outline',
  ),
  variant: {
    primary: cn(
      'bg-primary hover:bg-primary/90 text-primary-foreground',
      'disabled:border-primary/0 disabled:bg-primary/50',
    ),
    secondary: cn(
      'text-primary',
      'disabled:border-primary/50 disabled:text-primary/50',
      'bg-background [&:not(:disabled)]:hover:bg-foreground/5',
    ),
  },
};

const Button = ({ variant = 'primary', className, ...rest }: ButtonProps) => {
  return (
    <button
      type="button"
      className={cn(buttonClassNames.button, buttonClassNames.variant[variant], className)}
      {...rest}
    />
  );
};

const ButtonLink = forwardRef<HTMLAnchorElement, LinkButtonProps>(
  ({ className, variant = 'primary', ...rest }: LinkButtonProps, ref) => {
    return (
      <Link
        ref={ref}
        className={cn(buttonClassNames.button, buttonClassNames.variant[variant], className)}
        {...rest}
      />
    );
  },
);

export { Button, buttonClassNames, ButtonLink };
