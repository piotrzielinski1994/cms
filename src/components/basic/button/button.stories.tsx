import { DEFAULT_VALUE, getFallback } from '@/config/storybook/utils';
import type { Meta, StoryObj } from '@storybook/react';
import { useTranslations } from 'next-intl';
import { type ComponentProps } from 'react';
import { Button as ButtonComponent } from './button';

type Args = ComponentProps<typeof ButtonComponent> & {
  label?: string;
};

const meta: Meta<Args> = {
  component: ButtonComponent,
  title: 'Components/Basic/Button',
  argTypes: {
    label: { control: 'text' },
    variant: {
      control: 'select',
      options: ['primary', 'secondary'] satisfies Array<
        ComponentProps<typeof ButtonComponent>['variant']
      >,
    },
    disabled: { control: 'boolean' },
  },
  args: {
    label: DEFAULT_VALUE,
    variant: 'primary',
    disabled: false,
  },
};

const Render = ({ label, ...args }: Args) => {
  const t = useTranslations('storybook.basic.button');
  return (
    <div className="grid gap-4">
      <ButtonComponent {...args}>{getFallback(label, t('default'))}</ButtonComponent>
      <ButtonComponent disabled>{t('disabled')}</ButtonComponent>
      <ButtonComponent variant="secondary">{t('secondary')}</ButtonComponent>
      <ButtonComponent variant="secondary" disabled>
        {t('disabled')}
      </ButtonComponent>
    </div>
  );
};

const Button: StoryObj<typeof ButtonComponent> = { render: Render };

export { Button };
export default meta;
