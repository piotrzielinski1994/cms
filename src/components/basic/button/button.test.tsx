import { withProviders } from '@/utils/tests';
import { render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { describe, expect, it, vi } from 'vitest';
import { axe } from 'vitest-axe';
import { Button, ButtonLink } from './button';

describe('Button', () => {
  it('should have no accessibility violations', async () => {
    const { container } = render(<Button>Button</Button>);
    const results = await axe(container);
    expect(results).toHaveNoViolations();
  });

  it('should match the snapshot', () => {
    const { container } = render(<Button disabled>Button</Button>);
    expect(container).toMatchSnapshot();
  });

  it('calls onClick handler', async () => {
    const onClick = vi.fn();
    const { getByRole } = render(<Button onClick={onClick}>Button</Button>);

    await userEvent.click(getByRole('button', { name: 'Button' }));

    expect(onClick).toHaveBeenCalled();
  });
});

describe('ButtonLink', () => {
  it('renders with default primary variant classes', () => {
    const { getByRole } = render(withProviders()(<ButtonLink href="/test">ButtonLink</ButtonLink>));
    const link = getByRole('link');
    expect(link.className).toContain('bg-primary');
  });

  it('should have no accessibility violations', async () => {
    const { container } = render(withProviders()(<ButtonLink href="/test">ButtonLink</ButtonLink>));
    const results = await axe(container);
    expect(results).toHaveNoViolations();
  });

  it('should match the snapshot', () => {
    const { container } = render(withProviders()(<ButtonLink href="/test">ButtonLink</ButtonLink>));
    expect(container).toMatchSnapshot();
  });

  it('calls onClick handler', async () => {
    const onClick = vi.fn();
    const { getByRole } = render(
      withProviders()(
        <ButtonLink href="/test" onClick={onClick}>
          ButtonLink
        </ButtonLink>,
      ),
    );

    await userEvent.click(getByRole('link', { name: 'ButtonLink' }));

    expect(onClick).toHaveBeenCalled();
  });
});
