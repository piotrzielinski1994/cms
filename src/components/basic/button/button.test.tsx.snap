// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Button > should match the snapshot 1`] = `
<div>
  <button
    class="px-6 py-3 border-2 border-primary hover:border-primary/90 font-semibold whitespace-nowrap disabled:cursor-not-allowed focus-visible:tw-cms-outline bg-primary hover:bg-primary/90 text-primary-foreground disabled:border-primary/0 disabled:bg-primary/50"
    disabled=""
    type="button"
  >
    Button
  </button>
</div>
`;

exports[`ButtonLink > should match the snapshot 1`] = `
<div>
  <a
    class="px-6 py-3 border-2 border-primary hover:border-primary/90 font-semibold whitespace-nowrap disabled:cursor-not-allowed focus-visible:tw-cms-outline bg-primary hover:bg-primary/90 text-primary-foreground disabled:border-primary/0 disabled:bg-primary/50"
    href="/test"
  >
    ButtonLink
  </a>
</div>
`;
